text/0000-run-to-completion-async-fn.md

```diff
@@ -0,0 +1,879 @@
+also need to be aware that cancelling some of the current async functions (e.g.
+the proposed `scope` function for structured concurrency support in tokio) will
+will lead to `panic`s or `abort`s, since there exists no way to handle a
+cancellation in a reasonable fashion. By introducing run-to-completion function
+we will at least gain a feature on type-system level that will help to disambiguate
+functions which can safely be synchronously cancelled and ones which need to
+run to completion.
+
+### Additional verbosity
+
+The syntax `#[completion] async fn` is rather verbose. Since the requirement for
+completion functions will bubble up the call chain, it is expected that a lot
+of functions would need to carry the attribute. A more concise way to
+disambuigate the function types would be preferable, but might not be achievable
+within the rules of Rust 2018.
+
+However it is possible that IDE tools like rust-analyzer or Intellij-Rust could
+offer help - e.g. by auto-completing the attribute or rendering it in a more
+concise notation - in a similar fashion as IntelliJ renders some anonymous class
+instantiations in lambda syntax.
+
+### The new trait is `unsafe` by default
+
+The new proposed trait requires the implementation of an `unsafe` method. The
+`unsafe` annotation here is mainly required to enforce a special contract with
+the caller which is currently not achievable within Rusts safe type system.
+
+This is unfortunate, since we strive to minimize the use of `unsafe` code within
+Rust codebases. The `unsafe` annotation also allows implementors of the trait
+to use any arbitrary unsafe code, incl. code which could potentially lead to
+memory unsafety issues. This is not necessarily intended, because we want
+`unsafe` here mainly in order to enforce a contract with the caller.
+
+However we do not expect a lot of people to implement `RunToCompletionFuture`
+themselves. It should mainly be generated by the compiler. In addition this low
+level primitives like IO completion based socket types would need to be implemented
+by runtime authors - but those will need unsafe code anyway (they need to pass
+raw pointers to underlying C code or the kernel for an amount of time which can
+not be checked by lifetimes). Most users are purely expected to use high level
+`#[completion] async fn`s, which will not require unsafe code.
+
+We still expect people to write low level `Future`s by hand, for example to
+implement their custom channel type, timer, mutex, etc. However those types
+are typically supposed to be synchronously cancellable, and therefore people would continue to implement `Future`s instead of `RunToCompletionFuture`.
+the main exception here are future types which need to run to completion, like
+wrappers around IO completion operations. These are however unsafe by nature
+and require careful implementation and review. Therefore the `unsafe` attribute
+added here is justified.
```

Comment:
In fact, I think that unfortunately combinators are so deeply affected that the proposal becomes infeasible in its current state:

Say a combinator like select: once returned from it, what happens to the unfinished future? Return to caller just like today? Who ensures the future is polled again before dropped?

Same for FuturesUnordered: Poll a couple of items out, but leave some inside, then drop it - compiler is none the wiser -- and the invariant is broken.

**In general**: _A combinator that accepts completion futures **must poll all of its inner futures to completion before returning from any of its async functions**. It can neither store or return any incomplete futures because then they can be arbitrarily dropped._

Unfortunately, this gives us three bad options:

- Combinators disallow this - severely degrading their utility.
- Combinators allow this, but lose most of the guarantees of this proposal
- Drop-bombs - panic at runtime if futures and combinators are not polled to completion at drop-time.

It all boils down to this existing drop-semantics we have today - whether we're in async or not. Hence, some rudimentary linear type features are necessary, as far as I can tell.

Reply:
I hear you, but I think if we invalidate people's existing use of combinators, we're gonna have a harder time getting this through review. There's a lot more convincing to do.

FuturesUnordered is used under the hood in many other combinators and stream transforms too..

How would you solve the standard issue of:
- A TCP Listener produces a stream of Connections
- For each connection concurrently:
  - Run a handler which returns a result. The handler is a completion future.
- Collect the results out-of-order (in a buffer that never exceeds O(concurrent connections))

Reply:
We could use a completion version of `FuturesUnordered` that itself must be polled to completion.  Of course, there would need to be a method for cancellation as well.

Reply:
That is basically the structured concurrency `scope/TaskGroup/Nursery` thing.

https://github.com/tokio-rs/tokio/pull/2579 had an implementation for Rust/tokio without support for run to completion tasks. https://github.com/DougGregor/swift-evolution/blob/structured-concurrency/proposals/nnnn-structured-concurrency.md proposes a `TaskGroup` for Swift which is basically the same: A `FuturesUnordered` for run-to-completion coroutines with builtin support graceful cancellation.

And as said [on zulip](https://rust-lang.zulipchat.com/#narrow/stream/187312-wg-async-foundations/topic/run.20to.20completion.20async.20fn/near/222140636) I don't think the current `select!/join!` combinators should go away. They will still have their roles in waiting e.g. for a channel and a timer in parallel.

The new combinators will focus more on the orchestration of bigger tasks. E.g. on the server example that @betamos  had the TCP listener would open a task group, and start the processing of each connection as a subtask in this group. If the listener gets cancelled (e.g. via `SIGINT`), the subtasks also receive a graceful cancellations signal which still allows them to e.g. finish processing the current request, and then shut down.

Reply:
Btw: I think the original comment from @DemiMarie was about something different. I interpret it more has "combinators which can assume run to completion guarantees can be more efficient and offer different features", than a "old combinators are no longer useful".

I fully agree with that, but it's to be seen what combinators we really need. The number 1 on my list is really the `TaskGroup` thing, which however could also enable borrowing from the parent tasks as a new feature (like crossbeam scopes).